### PR TITLE
Zero distribution

### PR DESCRIPTION
**Pull request by tbrent**
Resolves https://github.com/code-423n4/2023-06-reserve-findings/issues/34

### CODE REVIEW

File: test/Revenues.test.ts [hunk 1]
```diff
@@ -636,6 +636,114 @@ describe(`Revenues - P${IMPLEMENTATION}`, () => {
...
+            emitted: true,
+          },
+          {
+            contract: token0,
+            name: 'Transfer',
+            args: [rsrTrader.address, backingManager.address, issueAmount.add(1)],
+            emitted: true,
+          },
+          {
+            contract: token1,
```
Comment (pmckelvy1) [inline test/Revenues.test.ts:685]: i dont think this should fire?

File: test/Revenues.test.ts [hunk 2]
```diff
@@ -636,6 +636,114 @@ describe(`Revenues - P${IMPLEMENTATION}`, () => {
...
+        ).to.be.revertedWith('rsrTotal > 0')
+        await distributor.setDistribution(STRSR_DEST, { rTokenDist: bn('0'), rsrDist: bn('0') })
+
+        // Should fail for unregistered token
+        await assetRegistry.connect(owner).unregister(collateral1.address)
+        await expect(
+          rsrTrader.returnTokens([rsr.address, token0.address, token1.address])
+        ).to.be.revertedWith('unregistered erc20')
+
+        // Succeed on just token1 + rsr
```
Comment (pmckelvy1) [inline test/Revenues.test.ts:670]: token0

File: test/Revenues.test.ts [hunk 3]
```diff
@@ -636,6 +636,114 @@ describe(`Revenues - P${IMPLEMENTATION}`, () => {
...
+            args: [rTokenTrader.address, backingManager.address, issueAmount],
+            emitted: true,
+          },
+          {
+            contract: token0,
+            name: 'Transfer',
+            args: [rTokenTrader.address, backingManager.address, issueAmount.add(1)],
+            emitted: true,
+          },
+          {
```
Comment (pmckelvy1) [inline test/Revenues.test.ts:738]: i don't think this should fire?

File: test/Revenues.test.ts [hunk 4]
```diff
@@ -636,6 +636,114 @@ describe(`Revenues - P${IMPLEMENTATION}`, () => {
...
+        ).to.be.revertedWith('rTokenTotal > 0')
+        await distributor.setDistribution(FURNACE_DEST, { rTokenDist: bn('0'), rsrDist: bn('0') })
+
+        // Should fail for unregistered token
+        await assetRegistry.connect(owner).unregister(collateral1.address)
+        await expect(
+          rTokenTrader.returnTokens([rsr.address, token0.address, token1.address])
+        ).to.be.revertedWith('unregistered erc20')
+
+        // Succeed on just token1 + rsr
```
Comment (pmckelvy1) [inline test/Revenues.test.ts:724]: token0

**Review by pmckelvy1 — CHANGES_REQUESTED** [review]

File: test/Revenues.test.ts (hunk 1 above)
Comment (tbrent) [reply test/Revenues.test.ts:685]: Agreed, it has `emitted: false` to check for that.

**Review by julianmrodri — COMMENTED** [review]
Looks good overall. Left some comments.

File: contracts/p0/RevenueTrader.sol [hunk 5]
```diff
@@ -50,6 +50,26 @@ contract RevenueTraderP0 is TradingP0, IRevenueTrader {
...
+    /// Return registered ERC20s to the BackingManager if distribution for tokenToBuy is 0
+    /// @custom:interaction
+    function returnTokens(IERC20[] memory erc20s) external notTradingPausedOrFrozen {
+        RevenueTotals memory revTotals = main.distributor().totals();
+        if (tokenToBuy == main.rsr()) {
+            require(revTotals.rsrTotal == 0, "rsrTotal > 0");
+        } else if (address(tokenToBuy) == address(main.rToken())) {
+            require(revTotals.rTokenTotal == 0, "rTokenTotal > 0");
+        } else {
+            revert("invalid tokenToBuy");
```
[the remark this answers is not in the document]
Comment (tbrent) [reply contracts/p0/RevenueTrader.sol:62]: Adding the comment

**Pull request merged**